cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +48,1538 @@ using parquet::detail::ColumnChunkDesc;
+               duration_cast<cudf::duration_ms>(cudf::duration_ns{nanos}).count();
+      case 1'000'000:  // microseconds
+        return duration_cast<cudf::duration_us>(duration_days).count() +
+               duration_cast<cudf::duration_us>(cudf::duration_ns{nanos}).count();
+      case 1'000'000'000:  // nanoseconds
+      default: return duration_cast<cudf::duration_ns>(duration_days).count() + nanos;
+    }
+  }();
+}
+
+/**
+ * @brief Helper function to convert a int64_t to a timestamp64
+ *
+ * @param value Value to convert
+ * @param timestamp_scale Timestamp scale
+ * @return Converted timestamp64 value
+ */
+__device__ __forceinline__ int64_t convert_to_timestamp64(int64_t const value,
+                                                          int32_t timestamp_scale)
+{
+  // Note: This function has been taken as-is from the snippet at:
+  // https://github.com/rapidsai/cudf/blob/c89c83c00c729a86c56570693b627f31408bc2c9/cpp/src/io/parquet/page_data.cuh#L247-L258
+
+  if (timestamp_scale < 0) {
+    // round towards negative infinity
+    int32_t const sign = (value < 0);
+    return ((value + sign) / -timestamp_scale) + sign;
+  } else {
+    return value * timestamp_scale;
+  }
+}
+
+/**
+ * @brief Query `cuco::static_set`s to evaluate (many) input (in)equality predicates
+ *
+ * @tparam Supported underlying data type of the cudf column
+ * @param decoded_data Span of storage for decoded values from all dictionaries
+ * @param results Span of device vector start pointers to store query results, one per predicate
+ * @param scalars Pointer to scalar device views, one per predicate
+ * @param operators Pointer to corresponding (in)equality operators, one per predicate
+ * @param set_storage Pointer to the start of the bulk cuco hash set slots
+ * @param set_offsets Pointer to offsets into the bulk set storage for each dictionary
+ * @param value_offsets Pointer to offsets into running sum of values in each dictionary
+ * @param total_row_groups Total number of row groups
+ * @param physical_type Parquet physical type of the column
+ */
+template <SupportedDictionaryType T>
+CUDF_KERNEL void query_dictionaries(cudf::device_span<T> decoded_data,
+                                    cudf::device_span<bool*> results,
+                                    ast::generic_scalar_device_view const* scalars,
+                                    ast::ast_operator const* operators,
+                                    slot_type* set_storage,
+                                    cudf::size_type const* set_offsets,
+                                    cudf::size_type const* value_offsets,
+                                    cudf::size_type total_row_groups,
+                                    parquet::Type physical_type)
+{
+  // Each thread block (cg) evaluates one scalar against all column chunk dictionaries (cuco hash
+  // sets) of this column
+  auto const group = cg::this_thread_block();
```

Comment:
What's the parallelization strategy here?  I'm not familiar with cuco, but why not just one thread (or one cg) per scalar per dictionary? Seems like this could go a lot wider.